### DISCUSSION TITLE
Interleaving Mapbox basemap labels with deckGL layers

### DISCUSSION THREAD

**Discussion by sethdumaguin** [visgl/deck.gl#5809] [Q&A]
Hello! We are trying to have the mapbox basemap labels be on top of all of the deckGL layers. We are following the sample found [here](https://deck.gl/docs/api-reference/mapbox/overview) I've added some code snippet of how we are doing it, but we are seeing some issues we've seen.

1. When our app loads and there are no layers added, the screen would start flashing until we add a layer

2. We use html2canvas to create a screenshot of our map, with this approach, the map is empty in the resulting screenshot

3. We are also using nebulaGL to draw on the map. With this approach of interleaving basemap labels and deckgl layer, we are unable to draw geometries on the map using nebula.

Appreciate any input y'all can give. Thank you

```javascript
const useInterleaveLayers = ({
  layers,
  mapGL,
  deckglMap,
  overlayOn,
}) => {
  const [mapLoaded, setMapLoaded] = useState(false);

  // List of layer ids that are already in the map
  const addedLayersRef = useRef([]);
  const map = mapGL?.getMap();
  const deck = deckglMap?.deck;

  useEffect(() => {
    if (map) {
      map.on('load', () => {
        setMapLoaded(true);
      });
    }
  }, [map]);

  if (!mapGL || !deckglMap || !mapLoaded) return;


  layers
    .filter((layer) => !addedLayersRef.current.includes(layer.id)) // Only add layers that are not in the mapbox gl yet
    .forEach((layer) => {
      map.addLayer(
        // This id has to match the id of the deck.gl layer
        new MapboxLayer({ id: layer.id, deck }),
        // Optionally define id from Mapbox layer stack under which to add deck layer
        overlayOn
      );
      addedLayersRef.current = [...addedLayersRef.current, layer.id];
    });
};

const App = ({ layers }) => {
  const { map, deckglMap } = ... ;
  useInterleaveLayers({
    layers: mapLayers,
    mapGL: map,
    deckglMap,
    overlayOn: activeOverlayOn, // id from Mapbox basemap where we want to overlay the layer on
  });

  return (
    <DeckGL
      viewState={...}
      onViewStateChange={...}
      layers={layers}
      ref={setDeckMap}
      onWebGLInitialized={setGLContext}
      glOptions={{
        /* To render vector tile polygons correctly */
        stencil: true,
      }}
    >
      {glContext && (
        <StaticMap
          ref={setMapRef}
          preserveDrawingBuffer
          gl={glContext}
        />
      )}
    </DeckGL>
  );
};
```

**Comment by Pessimistress** [comment]
It is a known issue with this usage pattern when there are no layers added - Deck is not aware that it is sharing context with a Mapbox map until at least one layer is added. The flashing is caused by Deck trying to wipe the context and render (nothing). You can add an empty layer at start. We should clarify this in the docs. I would greatly appreciate if you open an issue about this.

Screenshot: #4436 you need to either set `preserveDrawingBuffer` on Deck or capture the canvas inside `onAfterRender`. The latter is preferable as `preserveDrawingBuffer` has a perf cost.

You may need to ask in the nebula.gl repo but I don't think it's ever tested with the Mapbox custom layer integration.
